src/Fixer/PhpUnit/PhpUnitDedicateAssertFixer.php

```diff
@@ -45,13 +47,81 @@
         'is_null' => ['assertNotNull', 'assertNull'],
         'is_numeric' => true,
         'is_object' => true,
+        'is_readable' => ['assertNotIsReadable', 'assertIsReadable'],
         'is_real' => true,
         'is_resource' => true,
         'is_scalar' => true,
         'is_string' => true,
+        'is_writable' => ['assertNotIsWritable', 'assertIsWritable'],
     ];
 
     /**
+     * @var string[]
+     */
+    private $functions = [];
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure(array $configuration = null)
+    {
+        parent::configure($configuration);
+
+        if (isset($this->configuration['functions'])) {
```

Comment:
This should be done _before_ calling parent method so that you can check if `functions` option is actually passed. Then you would not need to allow `null` as its value.

Reply:
expected, new options: no `functions` option, only `target`.
then, if I pass config `[ target => 6.0 ]`, then `functions === null`. for that, I have to have `null` as possible value.

if I would remove `null` from possible value, that means I need to not call `parent::configure` method at all when `functions` is passed, so no validation of that array content at all...

Reply:
No: `functions` already has a default value, so it will always be defined after parent call.

But actually my comment was wrong: you don't need to change _when_ you do the `isset` check, but _what_ you check. Checking that `$configuration` has a `functions` key (instead of `$this->configuration`) will allow you to trigger the deprecation and know whether you can handle `target`.

Reply:
what is the difference of checking `$configuration` before parsing it and `$this->configuration` after parsing it?
`functions` has default value, which is... `null`, so how could I forbid `null` to be valid value ?

Reply:
`$configuration` is the raw configuration provided by the user, `$this->configuration` is the result of processing it (checking constraints and using defaults when needed).

The default value of `functions` being `null` is the result of the changes you made in the definition of the option. Previously, the option already had a default value, so not passing the option was already possible, you don't need to change that. What you need is to know whether the user explicitly passed `functions` option, hence my suggestion to check `$configuration` variable.

Reply:
`functions` on master has value, as value of it is used.
`functions` on my branch is nullable, as I want to remove that value in future, and in non-deprecated execution I want to have `functions` option empty.

Don't be distracted by internal property `functions` and option name `functions`, those are not the same.


so, to not walk around, please post a snippet what would you chance in both, `configure` method and `createConfigurationDefinition`

Reply:
```diff
diff --git a/README.rst b/README.rst
index 8921568c..decb4c7a 100644
--- a/README.rst
+++ b/README.rst
@@ -966,8 +966,12 @@ Choose from the list of available rules:
 
   Configuration options:
 
-  - ``functions`` (``null``): (deprecated, use ``target`` instead) List of assertions
-    to fix (overrides ``target``); defaults to ``null``
+  - ``functions`` (``array``): (deprecated, use ``target`` instead) List of assertions
+    to fix (overrides ``target``); defaults to ``['array_key_exists', 'empty',
+    'file_exists', 'is_array', 'is_bool', 'is_boolean', 'is_callable',
+    'is_double', 'is_float', 'is_infinite', 'is_int', 'is_integer',
+    'is_long', 'is_nan', 'is_null', 'is_numeric', 'is_object', 'is_real',
+    'is_resource', 'is_scalar', 'is_string']``
   - ``target`` (``'3.0'``, ``'3.5'``, ``'5.0'``, ``'5.6'``, ``'newest'``): target version of
     PHPUnit; defaults to ``'5.0'``
 
diff --git a/src/Fixer/PhpUnit/PhpUnitDedicateAssertFixer.php b/src/Fixer/PhpUnit/PhpUnitDedicateAssertFixer.php
index 08ae61a2..d318ea9f 100644
--- a/src/Fixer/PhpUnit/PhpUnitDedicateAssertFixer.php
+++ b/src/Fixer/PhpUnit/PhpUnitDedicateAssertFixer.php
@@ -67,8 +67,9 @@ public function configure(array $configuration = null)
     {
         parent::configure($configuration);
 
-        if (isset($this->configuration['functions'])) {
+        if (null !== $configuration && (isset($configuration['functions']) || isset($configuration[0]))) {
             @trigger_error('Option "functions" is deprecated and will be removed in 3.0, use option "target" instead.', E_USER_DEPRECATED);
             $this->functions = $this->configuration['functions'];
 
             return;
@@ -232,12 +233,11 @@ protected function createConfigurationDefinition()
 
         return new FixerConfigurationResolverRootless('functions', [
             (new FixerOptionBuilder('functions', '(deprecated, use `target` instead) List of assertions to fix (overrides `target`).'))
-                ->setAllowedTypes(['null', 'array'])
+                ->setAllowedTypes(['array'])
                 ->setAllowedValues([
-                    null,
                     (new FixerOptionValidatorGenerator())->allowedValueIsSubsetOf($values),
                 ])
-                ->setDefault(null)
+                ->setDefault($values)
                 ->getOption(),
             (new FixerOptionBuilder('target', 'Target version of PHPUnit.'))
                 ->setAllowedTypes(['string'])
```

Reply:
```
- ``functions`` (``array``): (deprecated, use ``target`` instead) List of assertions
  to fix (overrides ``target``); defaults to ``['array_key_exists', 'empty', ...
```

this is misleading to me, it's deprecated ,it overrides `target`, and then it has default value. so by defualt, it has value so it overrides target. that is what docs say.

and then, to handle that "overrides target", we need to
```diff
-        if (isset($this->configuration['functions'])) {
+        if (null !== $configuration && (isset($configuration['functions']) || isset($configuration[0]))) {
```

dont you say me this looks cleaner ;)

Reply:
I know, I initially reversed the logic to make `target` override `functions` instead. I then reverted that to provide a minimal diff. Though I think the new option should have precedence over the deprecated one when provided.

Reply:
but that would mean we either allow `target` to be nullable, or we don't, yet we bypass `FixerConfigurationResolver` and manage options manually... 
let us left it as-is for a now, yet I'm open for improving of this in future as apparently I don't see everything clearly here :(

Reply:
You wouldn't need to allow `target` to be `null` either. You just have to detect whether the user provides it explicitly. If we allow `null` in `functions` option here, we will not be able to revert it in the future without introducing deprecation handling, so I think we shouldn't introduce it at all.

Reply:
>  If we allow null in functions option here

that's the thing.
`functions` option is deprecated now and the only thing we will do with it is.... dropping it !

Reply:
Sure, but that doesn't mean we need to change its definition for that. By the way, as `null` would become allowed, one would be able to pass `['functions' => null]` without triggering the deprecation notice.

Reply:
good point, yet I don't think we handle it for any deprecated options currently

Reply:
I don't think we have any deprecated option that allows `null` either :p